Introduce v2 backend map with u32 backend ID

Weil0ng opened the pull request:
See commit msg.

Fixes: #16121

Downgrade is tracked in https://github.com/cilium/cilium/issues/17262

pchaigno reviewed (changes requested):
I don't understand why we still have references to the old map in this PR. Why are they needed? We can handle the downgrade by adding code to v1.10 for that (copy from v2 to v1).

> The 2/2 part will switch operation to v2 map completely

What is not already switched to v2 in this PR?

> introduce a --keep-legacy-service-backends flag to control whether we get rid of v1 map or keep it (for downgrade).

We usually just delete the maps after 3 releases, once nobody is supposed to be running a version with the old map.

Weil0ng committed:
This adds a new bpf map cilium_lb4_backends_v2

The V2 map key is u32 typed instead of u16.

Upon agent (re)start:
- we restore backends from v1 map and copy all
entries from v1 map to v2 map if v1 map exists
(this is the upgrade scenario).
- we then remove v1 map and operate on v2 map.

Signed-off-by: Weilong Cui <cuiwl@google.com>

pchaigno commented on bpf/lib/lb.h:
```diff
@@ -31,7 +31,16 @@ struct bpf_elf_map __section_maps LB6_SERVICES_MAP_V2 = {
+	.pinning        = PIN_GLOBAL_NS,
+	.max_elem       = CILIUM_LB_MAP_MAX_ENTRIES,
+	.flags          = CONDITIONAL_PREALLOC,
+};
```
Why do we still need the old map? I thought the goal was to switch maps in this PR.

Weil0ng replied:
> We usually just delete the maps after 3 releases, once nobody is supposed to be running a version with the old map.

Maybe I misunderstood you? Are you saying we don't explicitly "remove" existing v1 map until after 3 releases (but we do get rid of the code in this PR)?

pchaigno replied:
I'm saying we get rid of v1 now. It's the migration code that we can't remove until after 3 releases.

Weil0ng replied:
Makes sense, removed v1 map.

pchaigno commented on bpf/lib/lb.h:
```diff
@@ -29,9 +29,9 @@ struct bpf_elf_map __section_maps LB6_SERVICES_MAP_V2 = {
 };
 
-struct bpf_elf_map __section_maps LB6_BACKEND_MAP = {
+struct bpf_elf_map __section_maps LB6_BACKEND_MAP_V2 = {
```
The name of the macro could remain the same, no? Only its value needs to change.

Weil0ng replied:
True, but maybe we want to keep V2 in its name for better readability (in case we find a bug in the future due to this map key type change)? I see service map has a V2 in its name too :)

pchaigno commented on pkg/datapath/maps/map.go:
```diff
@@ -219,6 +221,9 @@ func (ms *MapSweeper) RemoveDisabledMaps() {
 	}
 
+	maps = append(maps, lbmap.Backend4MapName)
+	maps = append(maps, lbmap.Backend6MapName)
```
Are we sure the values have been migrated to the new map at this point? This function is usually for maps of disabled features. Maybe it would be best to remove the v1 backend maps in the same function that calls the migration logic?

Weil0ng replied:
Makes sense, done.

pchaigno commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -177,7 +177,7 @@ type FEPortName string
 
 // BackendID is the backend's ID.
-type BackendID uint16
+type BackendID = uint32
```
```suggestion
type BackendID uint32
```

Weil0ng replied:
I used type alias (with "=") on purpose to accommodate affinityMatch definition (where backendID is explicitly defined as uint32 type), but I think you are right, what I should do is change that definition to use lb.BackendID instead. Added that change, PTAL :)

pchaigno commented on pkg/maps/lbmap/maglev_outer_map.go:
```diff
@@ -104,7 +104,7 @@ func MaglevOuterMapTableSize(mapName string) (bool, uint32) {
 	defer innerMap.Close()
 
-	return true, innerMap.ValueSize() / uint32(unsafe.Sizeof(uint16(0)))
+	return true, innerMap.ValueSize() / uint32(unsafe.Sizeof(uint32(0)))
```
Does that mean we will also need to rename & migrate some of the Maglev maps?

Weil0ng replied:
hmm, good point, I think this will at least change the map size which seems to trigger a map regeneration? @brb

brb replied:
This will trigger the removal of the Maglev maps upon `lbmap.InitMaglevMaps()` invocation. The maps will be removed before the datapath has been regenerated, so for some time the service lookups for new connections will fail.

We could address this by temporarily renaming the old Maglev maps, and then removing it after the datapath reloading. The renaming needs to be conditional, and it would depend on v1->v2 is detected AND oldSize x 2 == newSize.

brb replied:
Ah, ignore what I wrote above. The map will be removed once we have removed all references to it (i.e., after reloading the datapath progs). So, all should be fine.

Weil0ng replied:
A bit confused, so this maglev map is not pinned, aka it's not suppose to survive across restarts?

brb replied:
The outer map is pinned, and the inner maps are referenced by the outer map, so they are not removed.

I've just tested the upgrade with Maglev enabled and `time.Sleep()` after the outer map has been removed - it works (the old maps are still there). However, after the upgrade, some requests gets dropped. This is because `LB{4,6}_MAGLEV_MAP_INNER` still has u16 instead of u32 as the val. This needs to be fixed in this PR.

Weil0ng replied:
Done. Although I still don't think I understand why we do not need to copy the maps (since value type change, wouldn't the bytes be offset to wrong position if we still access old map?)...is there a documentation describing the inner/outer map and how they work?

brb replied:
The new outer and inner maps will be created by `pkg/service` upon the startup, so no need to copy. We copy the backend maps though, and this is because we need to restore the backend ID allocator state. The new programs will reference the new outer map (and thus the new inner maps).

Weil0ng replied:
I see, so agent would be able to "recover" maglev hashing "state" without refering to the old map?

brb replied:
Yes.

Weil0ng replied:
Thanks, makes sense then, closing this thread since no more action to take (already updated inner map key type).

pchaigno commented on pkg/maps/lbmap/ipv4.go:
```diff
@@ -322,20 +348,37 @@ func (s *Service4Value) ToHost() ServiceValue {
 func NewBackend4Key(id loadbalancer.BackendID) *Backend4Key {
-	return &Backend4Key{ID: id}
+	return &Backend4Key{ID: uint16(id)}
 }
```
Do we still use this function, `NewBackend4Value`, and `NewBackend4`?

Weil0ng replied:
Ah good point, we no longer need these, removed.

pchaigno commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -225,12 +225,12 @@ func (*LBBPFMap) AddBackend(id uint16, ip net.IP, port uint16, ipv6 bool) error
 	}
 	if err != nil {
-		return fmt.Errorf("Unable to create backend (%d, %s, %d, %t): %s",
+		return fmt.Errorf("Unable to create backend v2 (%d, %s, %d, %t): %s",
```
Nit: I would drop this change. There is a single `AddBackend` method, so if we see this error message, we already know where it comes from and what it refers to :-)

Weil0ng replied:
Makes sense, removed :)

pchaigno commented on pkg/service/service.go:
```diff
@@ -219,6 +250,13 @@ func (s *Service) InitMaps(ipv6, ipv4, sockMaps, restore bool) error {
 	}
 
+	log.Infof("v1 be map exists: %v", v1BeMapExists)
+	if v1BeMapExists {
```
```suggestion
	if v1BeMapExists {
		log.Info("Backend map v1 exists. Migrating entries to backend map v2.")
```

Weil0ng replied:
Done.

pchaigno reviewed (commented):
Looks good to me overall. Maybe one concern around when we delete the v1 backend maps.

This will also definitely need a review from @brb, at least for the potential impact to the Maglev maps.

brb commented on pkg/service/service.go:
```diff
@@ -176,36 +178,83 @@ func (s *Service) GetCurrentTs() time.Time {
 	return time.Now()
 }
 
+func (s *Service) populateV2BeMapFromV1BeMap(ipv4, ipv6 bool) error {
```
Nit: in the past we tried to get rid of FE and BE terminology, so maybe populateBackendMapV2FromV1?

Weil0ng replied:
Done.

brb commented on pkg/service/service.go:
```diff
@@ -176,36 +178,83 @@ func (s *Service) GetCurrentTs() time.Time {
+			return fmt.Errorf("Unable to populate v2 lb4 backends map: %s", err)
+		}
+		lbmap.Backend4Map.Close()
+		removeMap(lbmap.Backend4MapName)
```
Maybe add a comment saying that the map will be actually removed once the last program referencing it has been removed.

Weil0ng replied:
It will be removed from the fs at this point, correct? So it would still be in-memory but if you do another restart, it should be gone completely? Just trying to understand, so who's referencing v1 map except for cilium-agent (which should not after restart?)?

brb replied:
The map is referenced by the BPF progs (in our case - bpf_host.o instances). Once the map is unpinned (=removed from the fs) and the last BPF prog referencing is gone, the map will be from memory.

Weil0ng replied:
I see, thanks! So bpf_host would be reloaded(removed then re-hooked) upon endpoint regeneration? Which endpoint hooks bpf_host though?

pchaigno replied:
The host endpoint (`reserved:host`) via `Loader.reloadHostDatapath`.

Weil0ng replied:
Ah I see, thanks!

brb reviewed (changes requested):
Thanks! I like the approach suggested by @pchaigno for the downgrade.

A few comments.

If you run the CI tests, the downgrade test case in the K8sUpdates suite should fail. So you will need to implement and backport the downgrade first.
